File: app/src/main/java/xyz/rodit/snapmod/features/tweaks/HideFriends.java
package xyz.rodit.snapmod.features.tweaks;

import android.text.TextUtils;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import de.robv.android.xposed.XC_MethodHook;
import xyz.rodit.snapmod.features.Feature;
import xyz.rodit.snapmod.features.FeatureContext;
import xyz.rodit.snapmod.mappings.ComposerFriend;
import xyz.rodit.snapmod.mappings.DisplayInfoContainer;
import xyz.rodit.snapmod.mappings.FriendListener;
import xyz.rodit.snapmod.mappings.ProfileMyFriendsSection;

public class HideFriends extends Feature {

    private final Set<String> hiddenFriends = new HashSet<>();

    public HideFriends(FeatureContext context) {
        super(context);
    }

    @Override
    protected void onConfigLoaded(boolean first) {
        hiddenFriends.clear();
        for (String username : context.config.getString("hidden_friends", "").split("\n")) {
            if (!TextUtils.isEmpty(username)) {
                hiddenFriends.add(username.trim());
            }
        }
    }

    @Override
    protected void performHooks() {
        // Hide friends from 'My Friends' in profile.
        ProfileMyFriendsSection.filter.hook(new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) {
                if (context.config.getBoolean("hide_friends")) {
                    List list = (List) param.args[0];
                    List filtered = new ArrayList();
                    for (Object o : list) {
                        String term = DisplayInfoContainer.wrap(o).getTerm();
                        if (!hiddenFriends.contains(term)) {
                            filtered.add(o);
                        }
                    }

                    param.args[0] = filtered;
                }
            }
        });

        // Hide friends from best friends list.
        FriendListener.handle.hook(new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) {
                if (context.config.getBoolean("hide_friends") && param.args[0] instanceof List) {
                    List list = (List) param.args[0];
                    if (list.isEmpty() || !ComposerFriend.isInstance(list.get(0))) {
                        return;
                    }

                    List filtered = new ArrayList();
                    for (Object o : list) {
                        String username = ComposerFriend.wrap(o).getUser().getUsername();
                        if (!hiddenFriends.contains(username)) {
                            filtered.add(o);
                        }
                    }

                    param.args[0] = filtered;
                }
            }
        });
    }
}

File: app/src/main/java/xyz/rodit/snapmod/features/tweaks/HideFriends.kt
package xyz.rodit.snapmod.features.tweaks
/*
import xyz.rodit.snapmod.features.Feature
import xyz.rodit.snapmod.features.FeatureContext
import xyz.rodit.snapmod.mappings.ComposerFriend
import xyz.rodit.snapmod.mappings.DisplayInfoContainer
import xyz.rodit.snapmod.mappings.FriendListener
import xyz.rodit.snapmod.mappings.ProfileMyFriendsSection
import xyz.rodit.snapmod.util.before

class HideFriends(context: FeatureContext) : Feature(context) {

    private val hiddenFriends: MutableSet<String> = HashSet()

    override fun onConfigLoaded(first: Boolean) {
        hiddenFriends.clear()
        for (username in context.config.getString("hidden_friends", "").split("\n")
            .filter { it.isNotBlank() }) {
            hiddenFriends.add(username.trim())
        }
    }

    override fun performHooks() {
        // Hide friends from 'My Friends' in profile.
        ProfileMyFriendsSection.filter.before(context, "hide_friends") {
            val list = it.args[0] as List<*>
            it.args[0] = list.filter { friend ->
                !hiddenFriends.contains(
                    DisplayInfoContainer.wrap(friend).term
                )
            }
        }

        // Hide friends from best friends list.
        FriendListener.handle.before(context, "hide_friends") {
            if (it.args[0] is List<*>) {
                val list = it.args[0] as List<*>
                if (list.isEmpty() || !ComposerFriend.isInstance(list[0])) {
                    return@before
                }

                it.args[0] =
                    list.filter { friend ->
                        !hiddenFriends.contains(
                            ComposerFriend.wrap(friend).user.username
                        )
                    }
            }
        }
    }
}
*/